File: goblinframework-core/src/main/java/org/goblinframework/core/config/ConfigParserManager.java
package org.goblinframework.core.config;

import org.goblinframework.api.config.ConfigParser;
import org.goblinframework.api.core.ThreadSafe;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.locks.ReentrantLock;

@ThreadSafe
final class ConfigParserManager {

  private final ReentrantLock lock = new ReentrantLock();
  private final List<ConfigParser> parsers = new ArrayList<>();

  ConfigParserManager() {
  }

  void register(@NotNull ConfigParser parser) {
    lock.lock();
    try {
      parsers.add(parser);
    } finally {
      lock.unlock();
    }
  }

  void parseConfigs() {
    List<ConfigParser> candidates;
    lock.lock();
    try {
      candidates = new ArrayList<>(parsers);
    } finally {
      lock.unlock();
    }
    candidates.parallelStream().forEach(ConfigParser::initialize);
  }
}

File: goblinframework-core/src/main/java/org/goblinframework/core/config/ConfigParserManager.kt
package org.goblinframework.core.config

import org.goblinframework.api.common.Install
import org.goblinframework.api.common.Singleton
import org.goblinframework.api.config.ConfigParser
import org.goblinframework.api.config.IConfigParserManager

@Singleton
class ConfigParserManager private constructor() : IConfigParserManager {

  companion object {
    @JvmField val INSTANCE = ConfigParserManager()
  }

  private val parsers = mutableListOf<ConfigParser>()

  @Synchronized
  override fun register(parser: ConfigParser) {
    parsers.add(parser)
  }

  fun asList(): List<ConfigParser> {
    return parsers.toMutableList()
  }

  @Install
  class Installer : IConfigParserManager by INSTANCE
}